File: packages/dotnet/src/AILabTools/AILabToolsClient.cs
using System;
using System.Net.Http;
using System.Threading;
using System.Threading.Tasks;
using AILabTools.Api;
using AILabTools.Exceptions;
using AILabTools.Generated.Params;
using AILabTools.Generated.Responses;
using AILabTools.Http;

namespace AILabTools;

/// <summary>Entry point for the asynchronous AILabTools C#/.NET SDK.</summary>
public sealed class AILabToolsClient : IDisposable
{
    private readonly HttpClient _httpClient;
    private readonly bool _ownsHttpClient;

    public AILabToolsClient(
        string apiKey,
        HttpClient? httpClient = null,
        string baseUrl = "https://www.ailabapi.com",
        TimeSpan? timeout = null)
    {
        if (string.IsNullOrWhiteSpace(apiKey)) throw new ArgumentException("AILabTools API key must not be empty.", nameof(apiKey));
        if (string.IsNullOrWhiteSpace(baseUrl)) throw new ArgumentException("Base URL must not be empty.", nameof(baseUrl));

        _ownsHttpClient = httpClient == null;
        _httpClient = httpClient ?? new HttpClient();
        if (timeout.HasValue) _httpClient.Timeout = timeout.Value;

        var requester = new Requester(_httpClient, apiKey, baseUrl);
        Common = new CommonApi(requester);
        Cutout = new CutoutApi(requester);
        Image = new ImageApi(requester);
        Portrait = new PortraitApi(requester);
        Background = new BackgroundApi(Cutout);
    }

    public CommonApi Common { get; }

    public CutoutApi Cutout { get; }

    public ImageApi Image { get; }

    public PortraitApi Portrait { get; }

    public BackgroundApi Background { get; }

    public async Task<CommonQueryAsyncTaskResultResponse> WaitForTaskAsync(
        string taskId,
        TimeSpan? interval = null,
        TimeSpan? timeout = null,
        bool throwOnFailed = true,
        CancellationToken cancellationToken = default)
    {
        if (string.IsNullOrWhiteSpace(taskId)) throw new ArgumentException("Task ID must not be empty.", nameof(taskId));
        var pollingInterval = interval ?? TimeSpan.FromSeconds(5);
        var pollingTimeout = timeout ?? TimeSpan.FromMinutes(5);
        var deadline = DateTimeOffset.UtcNow + pollingTimeout;

        while (DateTimeOffset.UtcNow <= deadline)
        {
            var result = await Common.CommonQueryAsyncTaskResultAsync(
                new CommonQueryAsyncTaskResultParams { TaskId = taskId },
                cancellationToken).ConfigureAwait(false);
            var status = result.Data?.TaskStatus ?? result.TaskStatus;
            if (status == 2) return result;
            if (status == 3)
            {
                if (throwOnFailed) throw new AILabToolsException("AILabTools task failed: " + taskId);
                return result;
            }
            await Task.Delay(pollingInterval, cancellationToken).ConfigureAwait(false);
        }

        throw new TimeoutException("AILabTools task timed out: " + taskId);
    }

    public void Dispose()
    {
        if (_ownsHttpClient) _httpClient.Dispose();
    }
}

File: packages/swift/Sources/AILabTools/AILabToolsClient.swift
import Foundation
#if canImport(FoundationNetworking)
import FoundationNetworking
#endif

public final class AILabToolsClient {
    public let common: CommonAPI
    public let cutout: CutoutAPI
    public let image: ImageAPI
    public let portrait: PortraitAPI
    public let background: BackgroundAPI

    private let session: URLSession
    private let ownsSession: Bool

    public init(
        apiKey: String,
        session: URLSession? = nil,
        baseURL: URL = URL(string: "https://www.ailabapi.com")!,
        timeout: TimeInterval = 30
    ) throws {
        guard !apiKey.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            throw AILabToolsError.invalidParameter("AILabTools API key must not be empty.")
        }
        if let session {
            self.session = session
            ownsSession = false
        } else {
            let configuration = URLSessionConfiguration.default
            configuration.timeoutIntervalForRequest = timeout
            self.session = URLSession(configuration: configuration)
            ownsSession = true
        }
        let requester = Requester(
            session: self.session,
            apiKey: apiKey,
            baseURL: baseURL,
            timeout: timeout
        )
        common = CommonAPI(requester: requester)
        cutout = CutoutAPI(requester: requester)
        image = ImageAPI(requester: requester)
        portrait = PortraitAPI(requester: requester)
        background = BackgroundAPI(cutout: cutout)
    }

    public func waitForTask(
        _ taskId: String,
        interval: TimeInterval = 5,
        timeout: TimeInterval = 300,
        throwOnFailure: Bool = true
    ) async throws -> CommonQueryAsyncTaskResultResponse {
        guard !taskId.isEmpty else {
            throw AILabToolsError.invalidParameter("Task ID must not be empty.")
        }
        let deadline = Date().addingTimeInterval(timeout)
        while Date() <= deadline {
            try Task.checkCancellation()
            let result = try await common.commonQueryAsyncTaskResult(
                CommonQueryAsyncTaskResultParams(taskId: taskId)
            )
            let status = result.data?.taskStatus ?? result.taskStatus
            if status == 2 { return result }
            if status == 3 {
                if throwOnFailure {
                    throw AILabToolsError.invalidResponse("AILabTools task failed: \(taskId)")
                }
                return result
            }
            let nanoseconds = UInt64(max(0, interval) * 1_000_000_000)
            try await Task.sleep(nanoseconds: nanoseconds)
        }
        throw AILabToolsError.invalidResponse("AILabTools task timed out: \(taskId)")
    }

    public func close() {
        if ownsSession {
            session.invalidateAndCancel()
        }
    }

    deinit {
        close()
    }
}
